### PR TITLE
Media category uploads

### PR DESCRIPTION
**Pull request by KTS915**
This PR supersedes PR #1414, which I will therefore close.

Here are some screenshots of what it makes happen:

Screenshot at 2024-08-24 15-28-50

Screenshot at 2024-08-24 15-29-22

Screenshot at 2024-08-24 15-30-37

Screenshot at 2024-08-24 15-31-07

The big difference with this PR is that it does not attempt to place a select dropdown within the uploader used on posts, pages, and widgets. The Javascript uploader library currently used by ClassicPress (Plupload) has no hooks available for such a purpose, and it causes any additional JavaScript to behave inconsistently. Instead, this PR simply adds a message about the current media upload category, together with a clickable link to the main Media library page, where the media category can be changed:

Screenshot at 2024-08-24 15-31-44

Note that this does not force uploads to use media categories. In fact, none of the additional features visible above will show on a site unless the use has gone to `Settings > Media` and selected `Organize uploads according to media category.`:

Screenshot at 2024-08-24 15-39-47

### CODE REVIEW

**Comment by mattyrob** [comment]
I've been working with this PR today and created some tests for the form output.

I've also been trying to get the Ajax functionality tested but so far have not got that working, but while working on this I suspect we need to add a `check_ajax_referer()` check within the `wp_ajax_media_cat_upload()` function.

**Comment by KTS915** [comment]
>I've also been trying to get the Ajax functionality tested but so far have not got that working, but while working on this I suspect we need to add a check_ajax_referer() check within the wp_ajax_media_cat_upload() function.

Doesn't adding the relevant action (here `media_cat_upload`) to the list in `admin-ajax.php` make such a check redundant? I thought that a nonce was necessary only when using one of the action hooks to receive the AJAX request (like `add_action( 'wp_ajax_new_action` or `add_action( 'wp_ajax_nopriv_new_action`).

**Comment by mattyrob** [comment]
> Doesn't adding the relevant action (here `media_cat_upload`) to the list in `admin-ajax.php` make such a check redundant? I thought that a nonce was necessary only when using one of the action hooks to receive the AJAX request (like `add_action( 'wp_ajax_new_action` or `add_action( 'wp_ajax_nopriv_new_action`).

I'm not 100% sure to be honest but many of the functions in `src/wp-admin/includes/ajax-actions.php` perform nonce checks before making changes to the database (like creating or deleting tags / comments etc, and as far as I can tell adding `media_cat_upload` in `src/wp-admin/admin-ajax.php` will dynamically create 'wp_ajax_media_cat_upload' and 'wp_ajax_nopriv_media_cat_upload' later in he code.

I'm wondering now if 'new_value' is too generic as a value name to pass also, perhaps that should be more specific to ensure plugins etc can't accidentally over-write core values.

**Review by xxsimoxx — APPROVED** [review]
Works for me.

I had to clear memcached cache, but I think this is normal.
Can't give an answer for the nonce checks.

**Comment by KTS915** [comment]
I'll try to make some time to work on both the `nonce` and variable name.

**Comment by mattyrob** [comment]
Screenshot 2024-09-05 at 08 43 13

I've just added a fix for this issue in unit testing.

**Comment by KTS915** [comment]
LGTM

**Comment by mattyrob** [comment]
> Actually, strike that. If I try to upload a file on the media grid page (i.e. `~/wp-admin/upload.php?mode=grid`), I'm now getting `An error occurred in the upload. Please try again later.`

I've just tested locally on Safari with the latest version of this branch is uploading from `wp-admin/upload.php?mode=grid` is working for me.

Was that fixed in your most recent commit?

**Comment by KTS915** [comment]
No, that one didn't fix it, but I think the commit I have just made has indeed fixed it.

I finally narrowed down the issue as occurring only after a change in the media upload category dropdown box had been made, and I believe that's because the code was then seeking to initialize the uploader when it was already initialized. So my latest commit first checks if the uploader has already been initialized.

**Comment by mattyrob** [comment]
@KTS915 - testing looks pretty good, one thing I've noticed though is that in grid view in the Media Library I am unable to change the Upload Media Category selector. I have one category for testing, and in list view and can change to the default of "Media Category" and see the error, but I cannot do the same in Grid view.

I can change from the default to my category in Grid view but now back again. Is that the expected behaviour?

**Comment by KTS915** [comment]
It is the expected.behavior, because the Pluploader that's used there has no mechanism for being turned on and off that we can hook into at that point. 

But you make a good point about the lack of a message. I'll see if I can add one later today.

**Comment by mattyrob** [comment]
> It is the expected.behavior, because the Pluploader that's used there has no mechanism for being turned on and off that we can hook into at that point.
> 
> But you make a good point about the lack of a message. I'll see if I can add one later today.

How about this as a fix - the option to have list or grid view is a user level database option, so in `src/wp-admin/includes/media.php` we can access that option with:

```
$mode = get_user_option( 'media_library_mode', get_current_user_id() ) ? get_user_option( 'media_library_mode', get_current_user_id() ) : 'grid';
```

Then use an if check when creating the drop down options:

```
if ( 'grid' !== $mode ) {
    $media_select .= '<option value=""' . selected( in_array( $cat_subfolder, array( '', '/' ) ), true, false ) . '>&nbsp;' . esc_html__( 'Media category' ) . '&nbsp;</option>';
}
```

**Comment by KTS915** [comment]
I have just submitted a couple of commits to add the message.

Sorry, @mattyrob , I'm only just seeing your latest message. But the problem with that solution is that it would only allow the changing of category in list mode. I don't think we need to be that restrictive. But this issue (and the previous, aborted PR) do highlight some of the issues with the PlUploader, which I'd love to replace at some point with something like Filepond.

**Comment by mattyrob** [comment]
After trying my change locally I was still able to change category in both List and Gird views. The only change was that dropdown in grid view did not contain the option to empty the value for category upload.

I just thought it might be a slightly more elegant way for user experience rather than present an option that cannot be used.

**Comment by KTS915** [comment]
OK, that makes sense!

How about undoing my last two commits, then, and apply your ideas instead?

**Comment by mattyrob** [comment]
@KTS915 - do you want me to do that?

**Comment by KTS915** [comment]
Yes, please!

**Comment by KTS915** [comment]
That looks great! I think we may finally have got this done!!

**Comment by xxsimoxx** [comment]
- In List View I can choose the upload media category, but I can't upload images from there. How should this be used?
- In list view I can choose "Media Category" as Media Category. Then, I get warned ("You need to choose a media category folder before you can upload a file."). Then I can switch to grid view, see the first dropdown element ad the Media Category, upload an image and the uploaded images has no Media category assigned. Is this expected?

**Comment by KTS915** [comment]
@xxsimoxx In List View, hit the Add New button to add new media.

**Comment by mattyrob** [comment]
> * In list view I can choose "Media Category" as Media Category. Then, I get warned ("You need to choose a media category folder before you can upload a file."). Then I can switch to grid view, see the first dropdown element ad the Media Category, upload an image and the uploaded images has no Media category assigned. Is this expected?

Should we change "Media Category" to "Select Media Category" to make this clearer - or even leave it empty?
When in Grid view, if the database doesn't have a default Media Category set, should we amend the dropdown to ask for a category to be selected? Or display a warning?

**Comment by xxsimoxx** [comment]
I think "Select Media Category is clearer", maybe this one can be disabled?
There is already a warning if the category is not selected.

**Comment by KTS915** [comment]
Latest commit changes the label to `Select Media Category`.

**Review by xxsimoxx — APPROVED** [review]
Great! Looks good to merge for me.
